packages/fast-components/src/anchored-region/anchored-region.ts

```diff
@@ -0,0 +1,978 @@
+     * the span in pixels of the selected position on each axis
+     */
+    private regionWidth: string;
+    private regionHeight: string;
+
+    private xTransformOrigin: string;
+    private yTransformOrigin: string;
+
+    private collisionDetector: IntersectionObserver;
+
+    private viewportRect: ClientRect | DOMRect | null;
+    private positionerDimension: Dimension;
+
+    private anchorTop: number;
+    private anchorRight: number;
+    private anchorBottom: number;
+    private anchorLeft: number;
+    private anchorHeight: number;
+    private anchorWidth: number;
+
+    private viewportScrollTop: number;
+    private viewportScrollLeft: number;
+
+    /**
+     * the positions currently being applied to layout
+     */
+    currentVerticalPosition: AnchoredRegionVerticalPositionLabel;
+    currentHorizontalPosition: AnchoredRegionHorizontalPositionLabel;
+
+    /**
+     * base offsets between the positioner's base position and the anchor's
+     */
+    private baseHorizontalOffset: number;
+    private baseVerticalOffset: number;
+
+    /**
+     * reference to the actual anchored container
+     */
+    public region: HTMLDivElement;
+
+    private openRequestAnimationFrame: number | null = null;
+
+    constructor() {
+        super();
+        this.handleAnchorResize = rafThrottle(this.handleAnchorResize);
+        this.handleRegionResize = rafThrottle(this.handleRegionResize);
+        this.setInitialState();
+    }
+
+    connectedCallback() {
+        super.connectedCallback();
+
+        if (isNil(this.viewportElement)) {
+            this.viewportElement = this.getViewport();
+        }
+
+        if (isNil(this.anchorElement)) {
+            this.anchorElement = this.getAnchor();
+        }
+
+        if (isNil(this.anchorElement) || isNil(this.viewportElement)) {
+            return;
+        }
+
+        this.connectObservers();
+
+        this.updateLayout();
+    }
+
+    public disconnectedCallback(): void {
+        super.disconnectedCallback();
+        this.disconnectObservers();
+    }
+
+    adoptedCallback() {
+        // TODO: do we need to handle this?
+    }
+
+    /**
+     * event thrown when the region's position changes
+     */
+    positionChanged() {
+        this.dispatchEvent(
+            new CustomEvent("changed", {
+                bubbles: true,
+                composed: true,
+            })
+        );
+    }
+
+    /**
+     * event thrown when the region's position changes
+     */
+    private updateLayoutForAttributeChange() {
+        if (this.initialLayoutComplete) {
+            this.requestLayoutUpdate();
+        }
+    }
+
+    /**
+     * resets the component
+     */
+    private reset() {
+        this.disconnectObservers();
+        this.setInitialState();
+        this.connectObservers();
+    }
+
+    /**
+     * sets the starting configuration for component internal values
+     */
+    private setInitialState = (): void => {
+        this.initialLayoutComplete = false;
```

Comment:
Any particular reason why you're not initializing these when they're being defined?

Reply:
Because we need to be able to reset the component to its initial state under certain circumstances, like when the anchor element changes, and it seemed best to define these in only one place.